prusti-viper/src/encoder/mir/pure/pure_functions/interface.rs

```diff
@@ -275,15 +276,19 @@ impl<'v, 'tcx: 'v> PureFunctionEncoderInterface<'v, 'tcx>
             );
 
             let maybe_identifier: SpannedEncodingResult<vir_poly::FunctionIdentifier> = (|| {
-                let (mut function, needs_patching) =
-                    if let Some(predicate_body) = self.get_predicate_body(proc_def_id, substs) {
-                        (
-                            pure_function_encoder.encode_predicate_function(&predicate_body)?,
-                            false,
-                        )
-                    } else if self.is_trusted(proc_def_id, Some(substs)) {
+                let (mut function, needs_patching) = match (
+                    self.is_trusted(proc_def_id, Some(substs)),
+                    self.get_proc_kind(proc_def_id, Some(substs)),
+                ) {
+                    // Predicate bodies are encoded even though the function is trusted
```

Comment:
This seems strange: do we need to support this? Can we not error out in parsing and/or not mark a predicate as trusted?

Reply:
My change there resulted from existing code, so I did not try to change existing behavior.
I double-checked: We can not mark a predicate as trusted.

Reply:
My question then is how should the encode work if we had trusted predicates? Should they be encoded as bodyless functions (like trusted pure functions) or should the body still be encoded? Before my change, a predicate would not have been encoded by a bodyless function.

Reply:
I guess they would be abstract/bodyless, yes… The motivation for predicates was to support Prusti syntax inside methods, but with a trusted method the body should be ignored anyway. So in the end, the only thing a trusted predicate does vs a trusted pure method is that it cannot be used in non-spec code. I guess that would be the case if you could declare these in an extern spec but maybe that's a stretch.

So, if we cannot mark a predicate as trusted anyway (not even via extern specs), can this code be simplified any (or at least the comment changed)?

Reply:
Yes, we also cannot add a `trusted` flag via extern spec, even though the error is not really helpful ("cannot use predicate in non-spec code" because of the expansion of the extern spec).

If the general behavior is to always encode a trusted function as bodyless I will adjust it.

Reply:
My naive understanding and refactoring caused `ui/predicate.rs` to fail.

I (maybe) understand this better now: Predicate functions *are* trusted. In contrast to pure functions, we cannot add a functional spec to the predicate, the predicate *is* the spec. Therefore, in the pure function encoder, we have to check first whether a function is a predicate to encode it, because they will otw. not do anything at all.

Does my explanation make sense? I think I can adjust the comment in the code a bit to clarify.

Reply:
This indirection of a spec item in a separate method is an implementation detail. I don't want to conflate this with `trusted` methods in the code or comments too much.